attest/signer.go

```diff
@@ -0,0 +1,128 @@
+package attest
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"encoding/asn1"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var tpm2HashFuncMap = map[crypto.Hash]tpm2.Algorithm{
+	crypto.SHA1:   tpm2.AlgSHA1,
+	crypto.SHA256: tpm2.AlgSHA256,
+	crypto.SHA384: tpm2.AlgSHA384,
+	crypto.SHA512: tpm2.AlgSHA512,
+}
+
+func tpm2HashFunc(h crypto.Hash) (tpm2.Algorithm, error) {
+	if a, ok := tpm2HashFuncMap[h]; ok {
+		return a, nil
+	}
+	return 0, fmt.Errorf("unsupported hash algorithm 0x%x", h)
+}
+
+type ec20Key struct {
+	tpm    io.ReadWriter
+	handle tpmutil.Handle
+	pub    crypto.PublicKey
+}
+
+type ecdsaSignature struct {
+	R, S *big.Int
+}
+
+func (k *ec20Key) Public() crypto.PublicKey {
+	return k.pub
+}
+
+func (k *ec20Key) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	hash, err := tpm2HashFunc(opts.HashFunc())
+	if err != nil {
+		return nil, err
+	}
+	scheme := &tpm2.SigScheme{Hash: hash, Alg: tpm2.AlgECDSA}
+	sig, err := tpm2.Sign(k.tpm, k.handle, akDefaultOwnerPassword, digest, scheme)
```

Comment:
TPM 2.0 Commands reference section 20.2.1 says:

```
If keyHandle references a restricted signing key, then validation shall be provided, indicating that the TPM
performed the hash of the data and validation shall indicate that hashed data did not start with
TPM_GENERATED_VALUE
```

The `tpm2.Sign` / `tpm2.SignWithSession` function [always uses a NULL ticket](https://github.com/google/go-tpm/blob/f0607eac7f8a894e8a26b848c072626f1cfe8b2d/tpm2/tpm2.go#L1328) with tag `TPM_ST_CHECKHASH` for the _validation_ argument, so the function in its current form will never work for AKs.

Reply:
@twitchy-jsonp great explanation. I would suggest @ericchiang to create new API tpm2.SignWithTicket instead of modifying or extending the existing ones. Just my 2 cents.

Reply:
This was taken over by https://github.com/google/go-attestation/pull/201 :) which does the right thing to create a crypto.Signer compatible interface